File: backend/src/models/Submission.ts
import { sequelize, DataTypes, Model } from '../config/database';
import { User } from './User';
import { Problem } from './Problem';

export class Submission extends Model {
  declare submissionId: number;
  declare userId: number;
  declare problemId: number;
  declare language: string;
  declare code: string;
  declare verdict: string;
  declare runtime: number;
}

Submission.init(
  {
    submissionId: {
      type: DataTypes.INTEGER,
      autoIncrement: true,
      primaryKey: true,
    },
    userId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: User,
        key: 'userId',
      },
    },
    problemId: {
      type: DataTypes.INTEGER,
      allowNull: false,
      references: {
        model: Problem,
        key: 'problemId',
      },
    },
    language: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    code: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
    verdict: {
      type: DataTypes.STRING,
      allowNull: false,
      defaultValue: 'PENDING',
    },
    runtime: {
      type: DataTypes.FLOAT,
      allowNull: true,
    },
  },
  {
    sequelize,
    tableName: 'submissions',
    timestamps: true,
    createdAt: 'submittedAt',
    updatedAt: false,
  }
);

User.hasMany(Submission, { foreignKey: 'userId', as: 'submissions' });
Submission.belongsTo(User, { foreignKey: 'userId', as: 'user' });

Problem.hasMany(Submission, { foreignKey: 'problemId', as: 'submissions' });
Submission.belongsTo(Problem, { foreignKey: 'problemId', as: 'problem' });

File: backend/src/main/java/com/codeforge/model/Submission.java
package com.codeforge.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import java.time.LocalDateTime;

/**
 * Submission entity — maps to the SUBMISSIONS table.
 * Records each code submission by a user for a problem.
 */
@Entity
@Table(name = "submissions")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Submission {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long submissionId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "problem_id", nullable = false)
    @JsonIgnoreProperties({"testCases", "hibernateLazyInitializer", "handler"})
    private Problem problem;

    @Column(nullable = false)
    private String language;

    @Column(columnDefinition = "TEXT", nullable = false)
    private String code;

    @Column(nullable = false)
    private String verdict; // PENDING, ACCEPTED, WRONG_ANSWER, TLE, COMPILATION_ERROR

    private Float runtime;

    @Column(name = "submitted_at", updatable = false)
    private LocalDateTime submittedAt = LocalDateTime.now();
}
